include/picongpu/fields/currentDeposition/EmZ/DepositCurrent.hpp

```diff
@@ -199,19 +220,26 @@ namespace picongpu
                     if(currentSurfaceDensityZ == 0.0_X)
                         return;
 
+                    auto const shape_i = make_TrajectoryAssignmentShapeFunction(
+                        typename T_Strategy::template Shape_i<ParticleAssign>(line.m_pos0[0], true),
+                        typename T_Strategy::template Shape_i<ParticleAssign>(line.m_pos1[0], true));
+
+                    auto const shape_j = make_TrajectoryAssignmentShapeFunction(
+                        typename T_Strategy::template Shape_j<ParticleAssign>(line.m_pos0[1], true),
+                        typename T_Strategy::template Shape_j<ParticleAssign>(line.m_pos1[1], true));
+
                     for(int j = T_begin; j < T_end; ++j)
                     {
-                        const float_X s0j = this->S0(line, j, 1);
-                        const float_X dsj = this->S1(line, j, 1) - s0j;
+                        const float_X s0j = shape_j.S0(j);
+                        const float_X dsj = shape_j.S1(j) - s0j;
                         for(int i = T_begin; i < T_end; ++i)
                         {
-                            const float_X s0i = this->S0(line, i, 0);
-                            const float_X dsi = this->S1(line, i, 0) - s0i;
-                            float_X W = s0i * this->S0(line, j, 1) + 0.5_X * (dsi * s0j + s0i * dsj)
-                                + (1.0_X / 3.0_X) * dsi * dsj;
+                            const float_X s0i = shape_i.S0(i);
+                            const float_X dsi = shape_i.S1(i) - s0i;
+                            float_X W = s0i * s0j + 0.5_X * (dsi * s0j + s0i * dsj) + (1.0_X / 3.0_X) * dsi * dsj;
```

Comment:
Unrelated to this PR, but in the spirit of separating this implementation from shape calculation details, it should use `auto` for these variables.